### ISSUE TITLE
Good design for encoders that generate several contexts

### ISSUE THREAD

**Issue by philip30** [neulab/xnmt#459]
Hi,

I want some encoder to generate multiple contexts for a single input.
DIrect usage for this that we allow the encoder to generate some samples from a single sentence and use all the samples for training.

Currently the implementation of multiple samples can be done in the decoder sides but not the encoder side. What is the best design for this?

**Comment by msperber** [comment]
If I understand correctly, this is a little similar to #445 where I added a CompoundInput. Would something like adding a CompoundExpressionSequence be suitable in your case?

As a remark, one thing I realized when doing that PR is that it's very helpful to given length methods an explicit name instead of using ```__len__```. For example, ```src.sent_len()```, ```src.batch_size()```, ```src.compound_len()``` are easier to understand and require less type checking than using ```len(src)``` which has a different meaning depending on the type of src.
